### PR TITLE
cmd: per command tls flags

### PR DESCRIPTION
**Pull request by runcom**
Based on #263 

@mtrmac @rhatdan PTAL

### CODE REVIEW

File: cmd/skopeo/utils.go [hunk 1]
```diff
@@ -9,12 +9,20 @@ import (
 	"github.com/urfave/cli"
 )
 
-func contextFromGlobalOptions(c *cli.Context, credsFlag string) (*types.SystemContext, error) {
+func contextFromGlobalOptions(c *cli.Context, credsFlag, certPathFlag, tlsVerifyFlag string) (*types.SystemContext, error) {
 	ctx := &types.SystemContext{
-		RegistriesDirPath:           c.GlobalString("registries.d"),
+		RegistriesDirPath: c.GlobalString("registries.d"),
+		// DEPRECATED: keep this here for backward compatibility, but override
+		// them if per subcommand flags are provided (see below).
```
Comment (runcom) [inline cmd/skopeo/utils.go]: This could entirely go away since we changed the way we lookup certs in `--cert-path` (no more `cert.pem`, `cert.key`), @mtrmac what do you think about getting rid of the global flags altogether?

File: cmd/skopeo/copy.go [hunk 2]
```diff
@@ -87,5 +87,23 @@ var copyCmd = cli.Command{
 			Value: "",
 			Usage: "Use `USERNAME[:PASSWORD]` for accessing the destination registry",
 		},
+		cli.StringFlag{
+			Name:  "src-cert-path",
```
Comment (rhatdan) [inline cmd/skopeo/copy.go]: Options are a little wordy,  -path adds nothing to this.

File: cmd/skopeo/copy.go [hunk 3]
```diff
@@ -87,5 +87,23 @@ var copyCmd = cli.Command{
...
+			Value: "",
+			Usage: "use certificates at `PATH` (*.crt, *.cert, *.key) to connect to the source registry",
+		},
+		cli.BoolTFlag{
+			Name:  "src-tls-verify",
+			Usage: "require HTTPS and verify certificates when talking to the docker source registry (defaults to true)",
+		},
+		cli.StringFlag{
+			Name:  "dest-cert-path",
```
Comment (rhatdan) [inline cmd/skopeo/copy.go]: Remove -path

File: cmd/skopeo/copy.go [hunk 4]
```diff
@@ -87,5 +87,23 @@ var copyCmd = cli.Command{
 			Value: "",
 			Usage: "Use `USERNAME[:PASSWORD]` for accessing the destination registry",
 		},
+		cli.StringFlag{
+			Name:  "src-cert-path",
+			Value: "",
+			Usage: "use certificates at `PATH` (*.crt, *.cert, *.key) to connect to the source registry",
+		},
+		cli.BoolTFlag{
+			Name:  "src-tls-verify",
```
Comment (rhatdan) [inline cmd/skopeo/copy.go]: Remove -verify

File: cmd/skopeo/copy.go [hunk 5]
```diff
@@ -87,5 +87,23 @@ var copyCmd = cli.Command{
...
+			Usage: "require HTTPS and verify certificates when talking to the docker source registry (defaults to true)",
+		},
+		cli.StringFlag{
+			Name:  "dest-cert-path",
+			Value: "",
+			Usage: "use certificates at `PATH` (*.crt, *.cert, *.key) to connect to the destination registry",
+		},
+		cli.BoolTFlag{
+			Name:  "dest-tls-verify",
```
Comment (rhatdan) [inline cmd/skopeo/copy.go]: remove -verify

File: cmd/skopeo/layers.go [hunk 6]
```diff
@@ -19,8 +19,25 @@ var layersCmd = cli.Command{
 	Name:      "layers",
 	Usage:     "Get layers of IMAGE-NAME",
 	ArgsUsage: "IMAGE-NAME [LAYER...]",
+	Hidden:    true,
```
Comment (mtrmac) [inline cmd/skopeo/layers.go]: This would be worth calling out… if not a separate PR, at least a separate commit?

File: cmd/skopeo/main.go [hunk 7]
```diff
@@ -31,13 +31,15 @@ func createApp() *cli.App {
...
+			Name:   "cert-path",
+			Value:  "",
+			Usage:  "use certificates at `PATH` (cert.pem, key.pem) to connect to the registry",
+			Hidden: true,
 		},
 		cli.BoolTFlag{
-			Name:  "tls-verify",
-			Usage: "require HTTPS and verify certificates when talking to docker registries (defaults to true)",
+			Name:   "tls-verify",
+			Usage:  "require HTTPS and verify certificates when talking to docker registries (defaults to true)",
```
Comment (mtrmac) [inline cmd/skopeo/main.go]: If these two are deprecated and don’t do anything, should this say so instead of describing what these options don’t do?  Or perhaps the `Usage` is never used for `Hidden` flags?

File: cmd/skopeo/main.go [hunk 8]
```diff
@@ -54,6 +56,12 @@ func createApp() *cli.App {
 		if c.GlobalBool("debug") {
 			logrus.SetLevel(logrus.DebugLevel)
 		}
+		if c.GlobalIsSet("cert-path") {
+			logrus.Warn("'--cert-path' is deprecated, please set this on the specific subcommand")
+		}
+		if c.GlobalIsSet("tls-verify") {
+			logrus.Warn("'--tls-verify' is deprecated, please set this on the specific subcommand")
```
Comment (mtrmac) [inline cmd/skopeo/main.go]: I am tempted to say that these cases should be a fatal error, same as any other unknown flag.

Comment (mtrmac) [inline cmd/skopeo/main.go]: … assuming that we would silently ignore them; as long as we do use the values, we should only warn of course.

File: cmd/skopeo/utils.go (hunk 1 above)
Comment (mtrmac) [inline cmd/skopeo/utils.go]: If `--cert-path` is not interoperable with the original flag, I don’t really see the point in introducing a new version which does something else but warns.

Looking at various pastebins I’ve seen so far, `--tls-verify=false` is used pretty frequently and keeping it working would be nice.

File: cmd/skopeo/utils.go [hunk 9]
```diff
@@ -9,12 +9,20 @@ import (
 	"github.com/urfave/cli"
 )
 
-func contextFromGlobalOptions(c *cli.Context, credsFlag string) (*types.SystemContext, error) {
+func contextFromGlobalOptions(c *cli.Context, credsFlag, certPathFlag, tlsVerifyFlag string) (*types.SystemContext, error) {
```
Comment (mtrmac) [inline cmd/skopeo/utils.go]: This is unwieldy… can this instead be a `flagPrefix string` (`""`, "src-"`, "dest-"`)?

File: cmd/skopeo/delete.go [hunk 10]
```diff
@@ -39,5 +39,14 @@ var deleteCmd = cli.Command{
...
 		},
+		cli.StringFlag{
+			Name:  "cert-path",
+			Value: "",
+			Usage: "use certificates at `PATH` (*.crt, *.cert, *.key) to connect to the registry",
+		},
+		cli.BoolTFlag{
+			Name:  "tls-verify",
+			Usage: "require HTTPS and verify certificates when talking to docker registries (defaults to true)",
+		},
```
Comment (mtrmac) [inline cmd/skopeo/delete.go]: Repeating this all over… could we have a `credsFlagList(prefix string, kind string) []cli.Flag)` in `utils.go` containing the canonical version of these flags, and the various subcommands just concatenating that with their own private flags?

(`kind` would be something like `""`/`"source"`/"destination"` for use in the `Usage:` fields.)

That way both the flag definitions and uses would be in a single place in `utils.go`.

File: cmd/skopeo/utils.go (hunk 1 above)
Comment (runcom) [inline cmd/skopeo/utils.go]: > > Looking at various pastebins I’ve seen so far, --tls-verify=false is used pretty frequently and keeping it working would be nice.

that means that if you use `skopeo copy` with just `--tls-verify` it'll be applied to both src and dest, is that ok with you?

**Comment by runcom** [comment]
> I assume you are certain enough that no existing user will be broken by this?

I assume so - and really, we're fixing something as copy should not send certs to both src and dest (I can keep compat with inspect and layers though)

File: cmd/skopeo/main.go (hunk 7 above)
Comment (runcom) [inline cmd/skopeo/main.go]: hidden doesn't show anything right

File: cmd/skopeo/utils.go (hunk 9 above)
Comment (runcom) [inline cmd/skopeo/utils.go]: sure

File: cmd/skopeo/delete.go (hunk 10 above)
Comment (runcom) [inline cmd/skopeo/delete.go]: fine

**Comment by runcom** [comment]
@mtrmac right now the main skopeo user is `atomic` - and it's not using `--cert-path` at all from what I can see - sure `--tls-verify` must be kept for backward compatibility there (assuming you're ok that a global `--tls-verify=true` as is right now, will switch off tls verification for src *AND* dest, as is right now :))

File: cmd/skopeo/main.go (hunk 8 above)
Comment (rhatdan) [inline cmd/skopeo/main.go]: I would go with Fatal.  Allowing the user to specify them and expecting something to happen, is not a good solution.  Especially for a security one like tls-verify.

File: cmd/skopeo/utils.go (hunk 1 above)
Comment (mtrmac) [inline cmd/skopeo/utils.go]: > that means that if you use `skopeo copy` with just `--tls-verify` it'll be applied to both src and dest,

That’s ugly, sure, but if this PR is already adding a warning about the flag being deprecated, it seems acceptable to me.

File: cmd/skopeo/main.go (hunk 8 above)
Comment (runcom) [inline cmd/skopeo/main.go]: I believe we can just Warn with `--tls-verify` because we still need to support that - I'll totally drop `--cert-path` though since hardly anyone is using that and there's no point in keeping backward compatibility.

File: cmd/skopeo/layers.go (hunk 6 above)
Comment (runcom) [inline cmd/skopeo/layers.go]: moved to a separate commit.

**Comment by runcom** [comment]
@mtrmac apart from https://github.com/projectatomic/skopeo/pull/264#discussion_r90896815 (which I'm not totally sure of) I fixed your comments, PTAL again

**Comment by mtrmac** [comment]
> apart from https://github.com/projectatomic/skopeo/pull/264#discussion_r90896815 (which I'm not totally sure of)

Yeah, splicing strings is ugly...

LGTM, make sure to drop the non-`master` c/i branch in `vendor.sh`
